sklearn/model_selection/_search_successive_halving.py

```diff
@@ -0,0 +1,898 @@
+from math import ceil, floor, log
+from abc import abstractmethod
+from numbers import Integral
+
+import numpy as np
+from ._search import _check_param_grid
+from ._search import BaseSearchCV
+from . import ParameterGrid, ParameterSampler
+from ..utils import check_random_state
+from ..utils.validation import _num_samples
+from ..base import is_classifier
+from ._split import check_cv
+from ..utils import resample
+
+
+__all__ = ['HalvingGridSearchCV', 'HalvingRandomSearchCV']
+
+
+class _TrainingFractionMetaSplitter:
+    """Splitter that subsamples the trainsets according to a given fraction"""
+    def __init__(self, *, base_cv, fraction, random_state):
+        self.base_cv = base_cv
+        self.fraction = fraction
+        self.random_state = check_random_state(random_state)
+
+    def split(self, X, y, groups=None):
+        for train_idx, test_idx in self.base_cv.split(X, y, groups):
+            train_idx = resample(
```

Comment:
if you want to follow the method you've had here until now, which changes the test sets in each iter, this resampling should be on all of X, i.e. outside of the loop.